build_tools/setup_helpers/extension.py

```diff
@@ -107,8 +109,36 @@ def _build_sentence_piece(debug):
     )
 
 
+def _build_re2(debug):
+    build_dir = _TP_BASE_DIR / 're2' / 'build_dir'
+    build_dir.mkdir(exist_ok=True)
+    build_env = os.environ.copy()
+    config = 'Debug' if debug else 'Release'
+    if platform.system() == 'Windows':
+        extra_args = ['-GNinja']
+        build_env.setdefault('CC', 'cl')
+        build_env.setdefault('CXX', 'cl')
+    else:
+        extra_args = []
+    subprocess.run(
+        args=['cmake', f'-DCMAKE_CXX_FLAGS=-fPIC', f'-DBUILD_SHARED_LIBS=OFF',
```

Comment:
Would you please try passing `-DCMAKE_CXX_FLAGS=/MT` for MSVC?

Reply:
Hi @peterjc123 

I tried that but somehow `/MD` is appended to the resulting compile command. The following is what I got with `-DCMAKE_EXPORT_COMPILE_COMMANDS=ON`.  `/MT` is passed but `/MD` is appended somehow. I checked `CMakeList.txt` and the resulting `Makefile` but there is no mention to `/MD` but somehow magically it shows up.

<Details><Summary> Generated commands </Summary></Details>

```json
[
{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\bitstate.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\bitstate.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/bitstate.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\compile.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\compile.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/compile.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\dfa.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\dfa.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/dfa.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\filtered_re2.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\filtered_re2.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/filtered_re2.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\mimics_pcre.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\mimics_pcre.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/mimics_pcre.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\nfa.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\nfa.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/nfa.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\onepass.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\onepass.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/onepass.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\parse.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\parse.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/parse.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\perl_groups.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\perl_groups.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/perl_groups.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\prefilter.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\prefilter.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/prefilter.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\prefilter_tree.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\prefilter_tree.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/prefilter_tree.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\prog.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\prog.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/prog.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\re2.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\thi
rd_party\\re2\\re2\\re2.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/re2.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\regexp.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\regexp.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/regexp.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\set.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\set.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/set.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\simplify.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\simplify.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/simplify.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\stringpiece.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\stringpiece.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/stringpiece.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\tostring.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\tostring.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/tostring.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\unicode_casefold.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\unicode_casefold.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/unicode_casefold.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\re2\\unicode_groups.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\re2\\unicode_groups.cc",
  "file": "C:/Users/circleci/project/third_party/re2/re2/unicode_groups.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\util\\rune.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\util\\rune.cc",
  "file": "C:/Users/circleci/project/third_party/re2/util/rune.cc"
},

{
  "directory": "C:/Users/circleci/project/third_party/re2/build_dir",
  "command": "C:\\PROGRA~2\\MICROS~2\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1426~1.288\\bin\\Hostx64\\x64\\cl.exe  /nologo /TP -DNOMINMAX -DSTRICT -DUNICODE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -I..\\  /MT /MD /O2 /Ob2 /DNDEBUG   /wd4100 /wd4201 /wd4456 /wd4457 /wd4702 /wd4815 /utf-8 /FoCMakeFiles\\re2.dir\\util\\strutil.cc.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\circleci\\project\\third_party\\re2\\util\\strutil.cc",
  "file": "C:/Users/circleci/project/third_party/re2/util/strutil.cc"
}
]
```

Reply:
This is an example build https://app.circleci.com/pipelines/github/pytorch/text/468/workflows/6a9574d2-f0f0-438b-9db1-54b36427bb6b/jobs/8164

I get errors like `re2.lib(dfa.cc.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MT_StaticRelease' in regex.obj`

Here is my attempts https://github.com/pytorch/text/pull/819/files .

I am very puzzled and trying to figure out where `/MD` is appended.

Reply:
I guess it is added by CMake on default. Usually we replace it with /MT when build_shared_libs=off when we write our own CMake script.

Reply:
https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html#variable:CMAKE_MSVC_RUNTIME_LIBRARY Looks like this may be useful.

Reply:
> https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html#variable:CMAKE_MSVC_RUNTIME_LIBRARY Looks like this may be useful.

I tried with `-DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded` but the command still has `/MD`.
I even tried to give invalid value as `-DCMAKE_MSVC_RUNTIME_LIBRARY=foo` and see if the first `cmake` would fail as the doc says. 

> `The value is ignored on non-MSVC compilers but an unsupported value will be rejected as an error when using a compiler targeting the MSVC ABI.`

but it did not raise error, so I am guessing that `CMAKE_MSVC_RUNTIME_LIBRARY` is not respected. 

> `Note This variable has effect only when policy CMP0091 is set to NEW prior to the first project() or enable_language() command that enables a language using a compiler targeting the MSVC ABI.`

This makes me wonder if we should add top-level `CMakeLists` and define the proper stuff there, then call the dependencies `CMakeLists` with `add_subdirectory`. Something similar to https://stackoverflow.com/q/35732776 . (I do not have much experience in CMake so not sure if this works or not.)

Reply:
Another option will be to apply patch to `CMakeLists.txt` in `RE2`.

Reply:
@mthrok Then I prefer the second one as it doesn't rely on CMake version. You could also create a top-level CMake script and then apply `/MD->/MT` changes base on the value of `BUILD_SHARED_LIBS` and finally you call `add_subdirectory` and it's all done.

Reply:
This is how we do that in PyTorch. https://github.com/pytorch/pytorch/blob/master/CMakeLists.txt#L307-L315